CONTRIBUTING.md

```diff
@@ -1,24 +1,134 @@
+| **Style guide**       | We use Checkstyle and Sonar to enforce beyond good practices.                                                                                                                                                                                                               |
+| **Exceptions**        | Specific exceptions live within utilities themselves and use `Exception` suffix e.g. `IdempotencyKeyException`.                                                                                                                                                             |
+| **Git commits**       | We follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). We do not enforce conventional commits on contributors to lower the entry bar. Instead, we enforce a conventional PR title so our label automation and changelog are generated correctly. |
+| **API documentation** | API reference docs are generated from Javadoc which should have examples to allow developers to have what they need within their own IDE. Documentation website covers the wider usage, tips, and strive to be concise.                                                     |
+| **Documentation**     | We treat it like a product. We sub-divide content aimed at getting started (80% of customers) vs advanced usage (20%). We also ensure customers know how to unit test their code when using our features.                                                                   |
+
+### Testing definition
+
+We group tests in different categories
+
+| Test              | When to write                                                                                         | Notes                                                                                                                           | Speed                                             |
+| ----------------- | ----------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------- |
+| Unit tests        | Verify the smallest possible unit works.                                                              | Networking access is prohibited. Prefer Functional tests given our complexity.                                                  | Lightning fast (nsec to ms)                       |
+| End-to-end tests  | Gain confidence that a Lambda function with our code operates as expected.                            | It simulates how customers configure, deploy, and run their Lambda function - Event Source configuration, IAM permissions, etc. | Slow (minutes)                                    |
+
+**NOTE**: Unit tests are mandatory. End-to-end tests should be created for new modules.
```

Comment:
Something here about test pyramid? Otherwise we'll end up with a test cube, or something 😢